File: packages/ai-engine/ai_engine/contracts/states.py
"""Business state enums mirror.

Single source of truth: `docs/contracts/state-machines.md` and
`packages/shared/src/states.ts`.
"""

from __future__ import annotations

from typing import Final, Literal

# ai_research_jobs.status — see docs/contracts/state-machines.md §1
AI_JOB_STATUS: Final[dict[str, str]] = {
    "QUEUED": "queued",
    "RUNNING": "running",
    "PARTIAL": "partial",
    "SUCCEEDED": "succeeded",
    "FAILED": "failed",
    "CANCELLED": "cancelled",
}
AiJobStatus = Literal["queued", "running", "partial", "succeeded", "failed", "cancelled"]

# ai_research_jobs.current_step — see docs/contracts/state-machines.md §2
AI_JOB_STEP: Final[dict[str, str]] = {
    "PLAN": "plan",
    "SEARCH": "search",
    "COMPRESS": "compress",
    "ANALYZE": "analyze",
    "WRITE": "write",
}
AiJobStep = Literal["plan", "search", "compress", "analyze", "write"]
# `current_step` always points to the LAST step that COMPLETED — see §2.
AI_JOB_STEP_ORDER: Final[tuple[AiJobStep, ...]] = (
    "plan",
    "search",
    "compress",
    "analyze",
    "write",
)

# content_import_jobs.status — see docs/contracts/state-machines.md §3
IMPORT_STATUS: Final[dict[str, str]] = {
    "QUEUED": "queued",
    "RUNNING": "running",
    "SUCCEEDED": "succeeded",
    "FAILED": "failed",
    "CANCELLED": "cancelled",
}
ImportStatus = Literal["queued", "running", "succeeded", "failed", "cancelled"]

# content_import_jobs.sourceKind — see docs/contracts/state-machines.md §11
IMPORT_SOURCE_KIND: Final[dict[str, str]] = {
    "FILE": "file",
    "CONFLUENCE": "confluence",
}
ImportSourceKind = Literal["file", "confluence"]

# summaries.status — see docs/contracts/state-machines.md §4
SUMMARY_STATUS: Final[dict[str, str]] = {
    "CANDIDATE": "candidate",
    "PENDING_REVIEW": "pending_review",
    "PUBLISHED": "published",
    "REJECTED": "rejected",
    "ARCHIVED": "archived",
}
SummaryStatus = Literal["candidate", "pending_review", "published", "rejected", "archived"]

# researches.status — see docs/contracts/state-machines.md §5
RESEARCH_STATUS: Final[dict[str, str]] = {
    "DRAFT": "draft",
    "PUBLISHED": "published",
    "ARCHIVED": "archived",
}
ResearchStatus = Literal["draft", "published", "archived"]

# researches.type
RESEARCH_TYPE: Final[dict[str, str]] = {
    "RESEARCH": "research",
    "KNOWLEDGE": "knowledge",
}
ResearchType = Literal["research", "knowledge"]

# researches.creation_method — see §7
CREATION_METHOD: Final[dict[str, str]] = {
    "MANUAL": "manual",
    "AI_RESEARCH": "ai_research",
    "FILE_IMPORT": "file_import",
    "CONFLUENCE_IMPORT": "confluence_import",
}
CreationMethod = Literal["manual", "ai_research", "file_import", "confluence_import"]

# comments.promote_status — see §8
PROMOTE_STATUS: Final[dict[str, str]] = {
    "NONE": "none",
    "NOMINATED": "nominated",
    "APPROVED": "approved",
    "REJECTED": "rejected",
}
PromoteStatus = Literal["none", "nominated", "approved", "rejected"]

# Source policy — see §9
SOURCE_POLICY: Final[dict[str, str]] = {
    "PREFER_USER_SOURCES": "prefer_user_sources",
    "ONLY_USER_SOURCES": "only_user_sources",
}
SourcePolicy = Literal["prefer_user_sources", "only_user_sources"]

# User role — see §10
USER_ROLE: Final[dict[str, str]] = {
    "MEMBER": "member",
    "ADMIN": "admin",
}
UserRole = Literal["member", "admin"]

# Week 6: AI 多轮追问会话状态（架构 §六点一）
AI_CHAT_SESSION_STATUS: Final[dict[str, str]] = {
    "ACTIVE": "active",
    "CLOSED": "closed",
}
AiChatSessionStatus = Literal["active", "closed"]

# Week 6: AI 多轮追问消息角色
AI_CHAT_ROLE: Final[dict[str, str]] = {
    "USER": "user",
    "ASSISTANT": "assistant",
}
AiChatRole = Literal["user", "assistant"]

# AI research job `reportType` (DB column is VarChar(40)). These values are
# persisted in the queue and rendered differently at the presentation boundary.
REPORT_TYPE: Final[dict[str, str]] = {
    "RESEARCH_REPORT": "research_report",
    "SUMMARY_BRIEF": "summary_brief",
    "SLIDES": "slides",
    "WEB_BRIEF": "web_brief",
    # Internal claim-scoped retrieval. It persists sources on the job but
    # must never create a second editable Research draft.
    "EVIDENCE_SEARCH": "evidence_search",
}
ReportType = Literal["research_report", "summary_brief", "slides", "web_brief", "evidence_search"]

# Helpers for partial-job rule (架构 §九 风险 10 / state-machines §1):
# mid-failure with at least 3 sources → partial; otherwise failed.
PARTIAL_MIN_SOURCES: Final[int] = 3


# ADR 0010: 研究目标 / 报告类型 / 专题 Issue
RESEARCH_OBJECTIVE: Final[dict[str, str]] = {
    "EXPLORE": "explore",
    "LEARN": "learn",
    "INVESTIGATE": "investigate",
    "DECIDE": "decide",
}
ResearchObjective = Literal[
    "explore", "learn", "investigate", "decide",
]

RESEARCH_OUTPUT_TYPE: Final[dict[str, str]] = {
    "MARKDOWN": "markdown",
    "SLIDES": "slides",
    "WEB": "web",
}
ResearchOutputType = Literal["markdown", "slides", "web"]

TOPIC_ISSUE_KIND: Final[dict[str, str]] = {
    "EVENT": "event",
    "PROBLEM": "problem",
}
TopicIssueKind = Literal["event", "problem"]

TOPIC_ISSUE_STATUS: Final[dict[str, str]] = {
    "ACTIVE": "active",
    "RESOLVED": "resolved",
    "ARCHIVED": "archived",
}
TopicIssueStatus = Literal["active", "resolved", "archived"]

RESEARCH_TOPIC_RELATION: Final[dict[str, str]] = {
    "AUTO": "auto",
    "MANUAL": "manual",
}
ResearchTopicRelation = Literal["auto", "manual"]

File: packages/shared/src/states.ts
// 业务状态机枚举。与 docs/contracts/state-machines.md 严格一致。

// ai_research_jobs.status
export const AI_JOB_STATUS = {
  QUEUED: 'queued',
  RUNNING: 'running',
  PARTIAL: 'partial',
  SUCCEEDED: 'succeeded',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
} as const;
export type AiJobStatus = (typeof AI_JOB_STATUS)[keyof typeof AI_JOB_STATUS];

// ai_research_jobs.current_step（架构 §十三 §3）
export const AI_JOB_STEP = {
  PLAN: 'plan',
  SEARCH: 'search',
  COMPRESS: 'compress',
  ANALYZE: 'analyze',
  WRITE: 'write',
} as const;
export type AiJobStep = (typeof AI_JOB_STEP)[keyof typeof AI_JOB_STEP];

// content_import_jobs.status（架构 §四点七）
export const IMPORT_STATUS = {
  QUEUED: 'queued',
  RUNNING: 'running',
  SUCCEEDED: 'succeeded',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
} as const;
export type ImportStatus = (typeof IMPORT_STATUS)[keyof typeof IMPORT_STATUS];

// summaries.status（架构 §五 + §四点七）
export const SUMMARY_STATUS = {
  CANDIDATE: 'candidate',
  PENDING_REVIEW: 'pending_review',
  PUBLISHED: 'published',
  REJECTED: 'rejected',
  ARCHIVED: 'archived',
} as const;
export type SummaryStatus = (typeof SUMMARY_STATUS)[keyof typeof SUMMARY_STATUS];

// researches.status（架构 §十二）
export const RESEARCH_STATUS = {
  DRAFT: 'draft',
  PUBLISHED: 'published',
  ARCHIVED: 'archived',
} as const;
export type ResearchStatus = (typeof RESEARCH_STATUS)[keyof typeof RESEARCH_STATUS];

// researches.type
export const RESEARCH_TYPE = {
  RESEARCH: 'research',
  KNOWLEDGE: 'knowledge',
} as const;
export type ResearchType = (typeof RESEARCH_TYPE)[keyof typeof RESEARCH_TYPE];

// researches.creation_method（架构 §五 + §四点七）
export const CREATION_METHOD = {
  MANUAL: 'manual',
  AI_RESEARCH: 'ai_research',
  FILE_IMPORT: 'file_import',
  CONFLUENCE_IMPORT: 'confluence_import',
} as const;
export type CreationMethod = (typeof CREATION_METHOD)[keyof typeof CREATION_METHOD];

// comments.promote_status（架构 §十四 + §十七）
export const PROMOTE_STATUS = {
  NONE: 'none',
  NOMINATED: 'nominated',
  APPROVED: 'approved',
  REJECTED: 'rejected',
} as const;
export type PromoteStatus = (typeof PROMOTE_STATUS)[keyof typeof PROMOTE_STATUS];

// AI 调研 source policy（架构 §四点五 + §九 风险 5b）
export const SOURCE_POLICY = {
  PREFER_USER_SOURCES: 'prefer_user_sources',
  ONLY_USER_SOURCES: 'only_user_sources',
} as const;
export type SourcePolicy = (typeof SOURCE_POLICY)[keyof typeof SOURCE_POLICY];

// 用户角色
export const USER_ROLE = {
  MEMBER: 'member',
  ADMIN: 'admin',
} as const;
export type UserRole = (typeof USER_ROLE)[keyof typeof USER_ROLE];

// Week 5 雷达：radar_sync_runs.status
export const RADAR_SYNC_STATUS = {
  RUNNING: 'running',
  COMPLETED: 'completed',
  PARTIAL: 'partial',
  FAILED: 'failed',
} as const;
export type RadarSyncStatus = (typeof RADAR_SYNC_STATUS)[keyof typeof RADAR_SYNC_STATUS];

// Week 5 雷达：radar_feedback.feedbackType
export const RADAR_FEEDBACK_TYPE = {
  USEFUL: 'useful',
  INACCURATE: 'inaccurate',
  USED: 'used',
  FAVORITE: 'favorite',
  SUGGEST_RESEARCH: 'suggest_research',
} as const;
export type RadarFeedbackType = (typeof RADAR_FEEDBACK_TYPE)[keyof typeof RADAR_FEEDBACK_TYPE];

// Week 5 雷达：候选处理状态（同步批次内）
export const RADAR_CANDIDATE_STATUS = {
  NEW: 'new',
  INTERPRETED: 'interpreted',
  SKIPPED: 'skipped',
  FAILED: 'failed',
} as const;
export type RadarCandidateStatus = (typeof RADAR_CANDIDATE_STATUS)[keyof typeof RADAR_CANDIDATE_STATUS];

// Week 6：AI 多轮追问会话状态（架构 §六点一）
export const AI_CHAT_SESSION_STATUS = {
  ACTIVE: 'active',
  CLOSED: 'closed',
} as const;
export type AiChatSessionStatus =
  (typeof AI_CHAT_SESSION_STATUS)[keyof typeof AI_CHAT_SESSION_STATUS];

// Week 6：AI 多轮追问消息角色（架构 §六点一）
export const AI_CHAT_ROLE = {
  USER: 'user',
  ASSISTANT: 'assistant',
} as const;
export type AiChatRole = (typeof AI_CHAT_ROLE)[keyof typeof AI_CHAT_ROLE];

// ADR 0010: 研究目标 / 报告类型 / 综述状态

export const RESEARCH_OBJECTIVE = {
  EXPLORE: 'explore',
  LEARN: 'learn',
  INVESTIGATE: 'investigate',
  DECIDE: 'decide',
} as const;
export type ResearchObjective =
  (typeof RESEARCH_OBJECTIVE)[keyof typeof RESEARCH_OBJECTIVE];

export const RESEARCH_OUTPUT_TYPE = {
  MARKDOWN: 'markdown',
  SLIDES: 'slides',
} as const;
export type ResearchOutputType =
  (typeof RESEARCH_OUTPUT_TYPE)[keyof typeof RESEARCH_OUTPUT_TYPE];

export const TOPIC_ISSUE_KIND = {
  EVENT: 'event',
  PROBLEM: 'problem',
} as const;
export type TopicIssueKind =
  (typeof TOPIC_ISSUE_KIND)[keyof typeof TOPIC_ISSUE_KIND];

export const TOPIC_ISSUE_STATUS = {
  ACTIVE: 'active',
  RESOLVED: 'resolved',
  ARCHIVED: 'archived',
} as const;
export type TopicIssueStatus =
  (typeof TOPIC_ISSUE_STATUS)[keyof typeof TOPIC_ISSUE_STATUS];

export const RESEARCH_TOPIC_RELATION = {
  AUTO: 'auto',
  MANUAL: 'manual',
} as const;
export type ResearchTopicRelation =
  (typeof RESEARCH_TOPIC_RELATION)[keyof typeof RESEARCH_TOPIC_RELATION];

// ADR 0010: 专题列表筛选
export const TOPIC_LIST_FILTER = {
  ALL: 'all',
  HOT: 'hot',
  WARMING: 'warming',
  EMERGING: 'emerging',
  FOLLOWED: 'followed',
} as const;
export type TopicListFilter =
  (typeof TOPIC_LIST_FILTER)[keyof typeof TOPIC_LIST_FILTER];
